flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/outlier/StochasticOutlierSelection.scala

```diff
@@ -0,0 +1,367 @@
+import org.apache.flink.api.common.operators.Order
+import org.apache.flink.api.common.typeinfo.TypeInformation
+import org.apache.flink.api.scala._
+import org.apache.flink.api.scala.utils._
+import org.apache.flink.ml.common.{LabeledVector, Parameter, ParameterMap, WithParameters}
+import org.apache.flink.ml.math.Breeze._
+import org.apache.flink.ml.math.{BreezeVectorConverter, Vector}
+import org.apache.flink.ml.pipeline.{TransformDataSetOperation, Transformer}
+
+import scala.language.implicitConversions
+import scala.reflect.ClassTag
+
+class StochasticOutlierSelection extends Transformer[StochasticOutlierSelection] {
+
+  import StochasticOutlierSelection._
+
+
+  /** Sets the perplexity of the outlier selection algorithm, can be seen as the k of kNN
+    * For more information, please read the Stochastic Outlier Selection algorithm paper
+    *
+    * @param perplexity the perplexity of the affinity fit
+    * @return
+    */
+  def setPerplexity(perplexity: Double): StochasticOutlierSelection = {
+    require(perplexity >= 1, "Perplexity must be at least one.")
+    parameters.add(Perplexity, perplexity)
+    this
+  }
+
+  /** The accepted error tolerance to save computational time when computing the affinity
+    *
+    * @param errorTolerance the accepted error tolerance with respect to the affinity
+    * @return
+    */
+  def setErrorTolerance(errorTolerance: Double): StochasticOutlierSelection = {
+    require(errorTolerance >= 0, "Error tolerance cannot be negative.")
+    parameters.add(ErrorTolerance, errorTolerance)
+    this
+  }
+
+  /** The maximum number of iterations to approximate the affinity
+    *
+    * @param maxIterations the maximum number of iterations
+    * @return
+    */
+  def setMaxIterations(maxIterations: Int): StochasticOutlierSelection = {
+    require(maxIterations > 0, "Maximum iterations must be positive.")
+    parameters.add(MaxIterations, maxIterations)
+    this
+  }
+
+}
+
+object StochasticOutlierSelection extends WithParameters {
+
+  // ========================================= Parameters ==========================================
+  case object Perplexity extends Parameter[Double] {
+    val defaultValue: Option[Double] = Some(30)
+  }
+
+  case object ErrorTolerance extends Parameter[Double] {
+    val defaultValue: Option[Double] = Some(1e-20)
+  }
+
+  case object MaxIterations extends Parameter[Int] {
+    val defaultValue: Option[Int] = Some(5000)
+  }
+
+  // ==================================== Factory methods ==========================================
+
+  def apply(): StochasticOutlierSelection = {
+    new StochasticOutlierSelection()
+  }
+
+  // ===================================== Operations ==============================================
+  case class BreezeLabeledVector(idx: Int, data: BreezeVector[Double])
+
+  implicit val transformLabeledVectors = {
+    new TransformDataSetOperation[StochasticOutlierSelection, LabeledVector, (Int, Double)] {
```

Comment:
Maybe we could add a comment that the label of `LabeledVector` should be the index or a unique integer value.